feat(preferred): add `bcrypt` replacements and docs (#642)
es-tooling/module-replacements: docs/modules/bcrypt.md

---
description: Modern alternatives to the bcrypt package for password hashing
---

# Replacements for `bcrypt`

> The native `bcrypt` package is a Node addon. Depending on your constraints, you may prefer a pure JavaScript implementation, or move to built-in cryptographic primitives.

Reviewer: ```suggestion
The `bcrypt` package can be replaced by native functionality in most runtimes, or more performant packages.
```

## `bcryptjs`

[`bcryptjs`](https://github.com/dcodeIO/bcrypt.js) is a widely used pure JavaScript implementation with a very similar API surface to `bcrypt`.

```ts
import bcrypt from 'bcrypt' // [!code --]
import bcrypt from 'bcryptjs' // [!code ++]

const salt = await bcrypt.genSalt(10) // [!code --]
const salt = await bcrypt.genSalt(10) // [!code ++]

const hash = await bcrypt.hash('password', salt) // [!code --]

> const hash = await bcrypt.hash('password', salt) // [!code ++]

Reviewer 2: Either

```suggestion
import bcrypt from 'bcryptjs' // [!code ++]

const salt = await bcrypt.genSalt(10)

const hash = await bcrypt.hash('password', salt)
```

or

```suggestion
import bcryptjs from 'bcryptjs' // [!code ++]

const salt = await bcrypt.genSalt(10) // [!code --]
const salt = await bcryptjs.genSalt(10) // [!code ++]

const hash = await bcrypt.hash('password', salt) // [!code --]
const hash = await bcryptjs.hash('password', salt) // [!code ++]
```

```

## `node:crypto` (native, Node.js built-in)

> Node provides [`node:crypto`](https://nodejs.org/api/crypto.html) for secure password hashing primitives (for example `scrypt` and `pbkdf2`). This is not a drop-in `bcrypt` API replacement, but it is a common “remove `bcrypt`” direction when you can standardize on a different KDF.

Reviewer: ```suggestion
Node provides [`node:crypto`](https://nodejs.org/api/crypto.html) for secure password hashing primitives (for example `scrypt` and `pbkdf2`). This is not a drop-in `bcrypt` API replacement, but is a good option if you can switch to a more secure cryptographic algorithm it supports.
```
